File: crates/mithril-oxide-sys/cpp/IR/BuiltinTypes.cpp
#include "BuiltinTypes.hpp"

#include <memory>

#include <mlir/IR/BuiltinAttributes.h>
#include <mlir/IR/TypeRange.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys {

const void* IntegerType_get(MLIRContext &context, unsigned int width, bool has_sign, bool is_signed)
{
    IntegerType::SignednessSemantics semantics;
    if(has_sign)
        semantics = is_signed
            ? IntegerType::SignednessSemantics::Signed
            : IntegerType::SignednessSemantics::Unsigned;
    else
        semantics = IntegerType::SignednessSemantics::Signless;

    return IntegerType::get(&context, width, semantics).getAsOpaquePointer();
}

const void* FunctionType_get(
    MLIRContext &context,
    rust::Slice<const void *const> inputs,
    rust::Slice<const void *const> results)
{
    std::vector<Type> inputs_vec;
    std::vector<Type> results_vec;

    for (const auto &value : inputs)
        inputs_vec.push_back(Type::getFromOpaquePointer(value));
    for (const auto &value : results)
        results_vec.push_back(Type::getFromOpaquePointer(value));

    return FunctionType::get(&context, inputs_vec, results_vec).getAsOpaquePointer();
}

#define MITHRIL_CAST_TO_SHAPED_TYPE_IMPL(FROM_TYPE) std::unique_ptr<ShapedType> FROM_TYPE ## _to_ShapedType(const FROM_TYPE &x) \
    { \
         return std::make_unique<ShapedType>(x); \
    }

MITHRIL_CAST_TO_SHAPED_TYPE_IMPL(TensorType);
MITHRIL_CAST_TO_SHAPED_TYPE_IMPL(RankedTensorType);
MITHRIL_CAST_TO_SHAPED_TYPE_IMPL(VectorType);
MITHRIL_CAST_TO_SHAPED_TYPE_IMPL(MemRefType);

} // namespace mithril_oxide_sys

File: crates/mithril-oxide-sys/src/IR/BuiltinTypes.rs
pub use ffi::{FunctionType_get, IntegerType_get};

#[cxx::bridge]
pub(crate) mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/BuiltinTypes.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/BuiltinTypes.hpp");

        type c_void = crate::c_void;

        // Constructors

        /// returns a type.
        #[must_use]
        pub fn IntegerType_get(
            context: Pin<&mut MLIRContext>,
            width: u32,
            has_sign: bool,
            is_signed: bool,
        ) -> *const c_void;

        /// # Safety
        /// - inputs: pointer to a valid Type
        /// - results: pointer to a valid Type
        /// - returns a type.
        #[must_use]
        pub unsafe fn FunctionType_get(
            context: Pin<&mut MLIRContext>,
            inputs: &[*const c_void],
            results: &[*const c_void],
        ) -> *const c_void;
    }
}

#[cfg(test)]
mod tests {
    use crate::IR::{MLIRContext::MLIRContext, Types::Type_print};

    use super::*;

    #[test]
    fn function_type() {
        let mut context = MLIRContext::new();

        unsafe {
            let a = IntegerType_get(context.pin_mut(), 1, false, false);
            let a_str = Type_print(a);
            assert_eq!("i1", a_str.as_str());

            let b = IntegerType_get(context.pin_mut(), 8, false, false);
            let b_str = Type_print(b);
            assert_eq!("i8", b_str.as_str());

            let func_type = FunctionType_get(context.pin_mut(), &[a], &[b]);
            let func_str = Type_print(func_type);
            assert_eq!("(i1) -> i8", func_str.as_str());
        }
    }
}
